File: level3/가장먼노드.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    vector<vector<bool>> graph(n,vector<bool>(n,false));
    for(int i = 0; i < edge.size(); i++){
        int x = edge[i][0], y = edge[i][1];
        graph[x-1][y-1] = true;
        graph[y-1][x-1] = true;
    }
    
    queue<int> q;
    q.push(0);
    vector<bool> visit(n,false);
    visit[0] = true;
    
    int answer = 0;
    while(!q.empty()){
        int q_size = q.size(), cnt = 0;
        for(int i = 0; i < q_size; i++){
            int vertex = q.front();
            q.pop();
            
            for(int i = 0; i < n; i++){
                if(graph[vertex][i] && !visit[i]){
                    cnt++;
                    q.push(i);
                    visit[i] = true;
                }
            }
        }     
        if(cnt != 0){
            answer = cnt;
        }
    }
    
    return answer;
}

File: level3/가장먼노드.py
import heapq

def solution(n, edge):
    vertex = {}
    for n1, n2 in edge:
        vertex[n1-1] = [n2-1] if n1-1 not in vertex.keys() else vertex[n1-1]+[n2-1]
        vertex[n2-1] = [n1-1] if n2-1 not in vertex.keys() else vertex[n2-1]+[n1-1]            
        
    dist = [float('inf')]*n
    dist[0] = 0
    
    q = []
    heapq.heappush(q,(0,0))
    while q:
        d, n = heapq.heappop(q)
        for i in vertex[n]:
            if dist[i] > d+1:
                dist[i] = d+1
                heapq.heappush(q,(d+1,i))
    
    max_dist = 0
    answer = 0
    for d in dist:
        if d > max_dist:
            max_dist = d
            answer = 1
        elif d == max_dist:
            answer += 1
            
    return answer
